qcodes/instrument/parameter.py

```diff
@@ -929,6 +939,126 @@ def sweep(self, start, stop, step=None, num=None):
+        for setpointarray in setpoints:
+            if not isinstance(setpointarray, _BaseParameter):
+                raise TypeError(f"Setpoints is of type {type(setpointarray)}"
+                                f" expcected a QCoDeS parameter")
+        self._setpoints = setpoints
+
+    def validate_consistent_shape(self) -> None:
+        """
+        Verifies that the shape of the Array Validator of the parameter
+        is consistent with the Validator of the Setpoints. This requires that
+        both the setpoints and the actual parameters have validators
+        of type Arrays with a defined shape.
+        """
+
+        if not isinstance(self.vals, Arrays):
+            raise ValueError(f"Can only validate shapes for parameters "
+                             f"with Arrays validator. {self.name} does "
+                             f"not have an Arrays validator.")
+        output_shape = self.vals.shape
+        setpoints_shape_list: List[Optional[Union[int, Callable[[], int]]]] = []
+        for sp in self.setpoints:
+            if not isinstance(sp.vals, Arrays):
+                raise ValueError(f"Can only validate shapes for parameters "
+                                 f"with Arrays validator. {sp.name} is "
+                                 f"a setpoint vector but does not have an "
+                                 f"Arrays validator")
+            if sp.vals.shape is not None:
+                setpoints_shape_list.extend(sp.vals.shape)
+            else:
+                setpoints_shape_list.append(sp.vals.shape)
+        setpoints_shape = tuple(setpoints_shape_list)
+
+        if output_shape is None:
+            raise ValueError(f"Trying to validate shape but parameter "
+                             f"{self.name} does not define a shape")
+        if None in output_shape or None in setpoints_shape:
+            raise ValueError(f"One or more dimensions have unknown shape "
+                             f"when comparing output: {output_shape} to "
+                             f"setpoints: {setpoints_shape}")
+
+        if output_shape != setpoints_shape:
+            raise ValueError(f"Shape of output is not consistent with "
+                             f"setpoints. Output is shape {output_shape} and "
+                             f"setpoints are shape {setpoints_shape}")
+        log.info(f"For parameter {self.full_name} verified "
+                 f"that {output_shape} matches {setpoints_shape}")
+
+    def validate(self, value: ParamDataType) -> None:
+        """
+        Overwrites the standard `validate` to also check the
+        the parameter has consistent shape with it's setpoints.
+        This only makes sense if the parameter has an Arrays validator
+
+        Arguments are passed to the super method
+        """
+        if isinstance(self.vals, Arrays):
+            self.validate_consistent_shape()
+        super().validate(value)
+
+
+class GeneratedSetPoints(Parameter):
```

Comment:
As William suggested perhaps this is not general enough and should be left to implement when needed.